// @redux-requests/react/lib/mutation.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactRedux = require("react-redux");

var _core = require("@redux-requests/core");

var _propTypesValidators = require("./propTypesValidators");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Mutation = function Mutation(_ref) {
  var type = _ref.type,
      selector = _ref.selector,
      requestKey = _ref.requestKey,
      children = _ref.children,
      Component = _ref.component,
      extraProps = _objectWithoutPropertiesLoose(_ref, ["type", "selector", "requestKey", "children", "component"]);

  var mutation = (0, _reactRedux.useSelector)(selector || (0, _core.getMutationSelector)({
    type: type,
    requestKey: requestKey
  }));

  if (children) {
    return children(mutation);
  }

  return /*#__PURE__*/_react["default"].createElement(Component, _extends({
    mutation: mutation
  }, extraProps));
};

Mutation.propTypes = process.env.NODE_ENV !== "production" ? {
  selector: _propTypes["default"].func,
  type: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].func]),
  requestKey: _propTypes["default"].string,
  children: _propTypes["default"].func,
  component: (0, _propTypesValidators.reactComponentPropType)('Mutation')
} : {};
var _default = Mutation;
exports["default"] = _default;
