// @redux-requests/react/lib/use-query.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = require("react");

var _reactRedux = require("react-redux");

var _core = require("@redux-requests/core");

var _useDispatchRequest = _interopRequireDefault(require("./use-dispatch-request"));

var _requestsContext = _interopRequireDefault(require("./requests-context"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var emptyVariables = [];

var useQuery = function useQuery(_ref) {
  var _ref$variables = _ref.variables,
      variables = _ref$variables === void 0 ? emptyVariables : _ref$variables,
      autoLoad = _ref.autoLoad,
      autoReset = _ref.autoReset,
      throwError = _ref.throwError,
      suspense = _ref.suspense,
      suspenseSsr = _ref.suspenseSsr,
      selectorProps = _objectWithoutPropertiesLoose(_ref, ["variables", "autoLoad", "autoReset", "throwError", "suspense", "suspenseSsr"]);

  var requestContext = (0, _react.useContext)(_requestsContext["default"]);
  suspenseSsr = suspenseSsr === undefined ? requestContext.suspenseSsr : suspenseSsr;
  suspense = suspense === undefined ? requestContext.suspense : suspense;
  autoLoad = autoLoad === undefined ? requestContext.autoLoad : autoLoad;
  autoReset = autoReset === undefined ? requestContext.autoReset : autoReset;
  throwError = throwError === undefined ? requestContext.throwError : throwError;
  var dispatchRequest = (0, _useDispatchRequest["default"])();
  var store = (0, _reactRedux.useStore)();
  var key = "" + selectorProps.type + (selectorProps.requestKey || '');
  var dispatchQuery = (0, _react.useCallback)(function () {
    return dispatchRequest((selectorProps.action || selectorProps.type).apply(void 0, variables));
  }, [selectorProps.action, selectorProps.type].concat(variables));
  var dispatchStopPolling = (0, _react.useCallback)(function () {
    dispatchRequest((0, _core.stopPolling)([{
      requestType: selectorProps.type,
      requestKey: selectorProps.requestKey
    }]));
  }, [selectorProps.type, selectorProps.requestKey]);
  (0, _react.useEffect)(function () {
    if (autoLoad) {
      dispatchQuery();
    }
  }, [autoLoad, dispatchQuery]);
  var query = (0, _reactRedux.useSelector)((0, _core.getQuerySelector)(selectorProps));
  (0, _react.useEffect)(function () {
    dispatchRequest((0, _core.addWatcher)(key));
    return function () {
      dispatchRequest((0, _core.removeWatcher)(key));

      if (autoReset && !store.getState().requests.watchers[key]) {
        dispatchRequest((0, _core.resetRequests)([{
          requestType: selectorProps.type,
          requestKey: selectorProps.requestKey
        }], true, false));
      }
    };
  }, [selectorProps.type, selectorProps.requestKey]);

  if (suspenseSsr && (query.loading || query.pristine)) {
    if (autoLoad && query.pristine) {
      throw dispatchQuery();
    }

    throw dispatchRequest((0, _core.joinRequest)(key, autoLoad));
  }

  if (suspense && !suspenseSsr && query.loading) {
    throw dispatchRequest((0, _core.joinRequest)(key));
  }

  if (throwError && query.error) {
    throw {
      error: query.error,
      type: selectorProps.type,
      requestKey: selectorProps.requestKey
    };
  }

  return (0, _react.useMemo)(function () {
    return _extends({}, query, {
      load: dispatchQuery,
      stopPolling: dispatchStopPolling
    });
  }, [query, dispatchQuery, dispatchStopPolling]);
};

var _default = useQuery;
exports["default"] = _default;
