// @redux-requests/core/lib/constants.js
"use strict";

exports.__esModule = true;
exports.STOP_SUBSCRIPTIONS = exports.CLOSE_WEBSOCKET = exports.OPEN_WEBSOCKET = exports.GET_WEBSOCKET = exports.WEBSOCKET_CLOSED = exports.WEBSOCKET_OPENED = exports.STOP_POLLING = exports.JOIN_REQUEST = exports.REMOVE_WATCHER = exports.ADD_WATCHER = exports.SET_UPLOAD_PROGRESS = exports.SET_DOWNLOAD_PROGRESS = exports.ABORT_REQUESTS = exports.RESET_REQUESTS = exports.CLEAR_REQUESTS_CACHE = exports.ABORT_SUFFIX = exports.ERROR_SUFFIX = exports.SUCCESS_SUFFIX = void 0;
var SUCCESS_SUFFIX = '_SUCCESS';
exports.SUCCESS_SUFFIX = SUCCESS_SUFFIX;
var ERROR_SUFFIX = '_ERROR';
exports.ERROR_SUFFIX = ERROR_SUFFIX;
var ABORT_SUFFIX = '_ABORT';
exports.ABORT_SUFFIX = ABORT_SUFFIX;
var CLEAR_REQUESTS_CACHE = 'CLEAR_REQUESTS_CACHE';
exports.CLEAR_REQUESTS_CACHE = CLEAR_REQUESTS_CACHE;
var RESET_REQUESTS = 'RESET_REQUESTS';
exports.RESET_REQUESTS = RESET_REQUESTS;
var ABORT_REQUESTS = 'ABORT_REQUESTS';
exports.ABORT_REQUESTS = ABORT_REQUESTS;
var SET_DOWNLOAD_PROGRESS = 'SET_DOWNLOAD_PROGRESS';
exports.SET_DOWNLOAD_PROGRESS = SET_DOWNLOAD_PROGRESS;
var SET_UPLOAD_PROGRESS = 'SET_UPLOAD_PROGRESS';
exports.SET_UPLOAD_PROGRESS = SET_UPLOAD_PROGRESS;
var ADD_WATCHER = 'ADD_WATCHER';
exports.ADD_WATCHER = ADD_WATCHER;
var REMOVE_WATCHER = 'REMOVE_WATCHER';
exports.REMOVE_WATCHER = REMOVE_WATCHER;
var JOIN_REQUEST = 'JOIN_REQUEST';
exports.JOIN_REQUEST = JOIN_REQUEST;
var STOP_POLLING = 'STOP_POLLING';
exports.STOP_POLLING = STOP_POLLING;
var WEBSOCKET_OPENED = 'WEBSOCKET_OPENED';
exports.WEBSOCKET_OPENED = WEBSOCKET_OPENED;
var WEBSOCKET_CLOSED = 'WEBSOCKET_CLOSED';
exports.WEBSOCKET_CLOSED = WEBSOCKET_CLOSED;
var GET_WEBSOCKET = 'GET_WEBSOCKET';
exports.GET_WEBSOCKET = GET_WEBSOCKET;
var OPEN_WEBSOCKET = 'OPEN_WEBSOCKET';
exports.OPEN_WEBSOCKET = OPEN_WEBSOCKET;
var CLOSE_WEBSOCKET = 'CLOSE_WEBSOCKET';
exports.CLOSE_WEBSOCKET = CLOSE_WEBSOCKET;
var STOP_SUBSCRIPTIONS = 'STOP_SUBSCRIPTIONS';
exports.STOP_SUBSCRIPTIONS = STOP_SUBSCRIPTIONS;
