// @redux-requests/react/es/query.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { getQuerySelector } from '@redux-requests/core';
import { reactComponentPropType } from './propTypesValidators';

var Query = function Query(_ref) {
  var type = _ref.type,
      requestKey = _ref.requestKey,
      selector = _ref.selector,
      defaultData = _ref.defaultData,
      multiple = _ref.multiple,
      children = _ref.children,
      Component = _ref.component,
      isDataEmpty = _ref.isDataEmpty,
      showLoaderDuringRefetch = _ref.showLoaderDuringRefetch,
      noDataMessage = _ref.noDataMessage,
      ErrorComponent = _ref.errorComponent,
      errorComponentProps = _ref.errorComponentProps,
      LoadingComponent = _ref.loadingComponent,
      loadingComponentProps = _ref.loadingComponentProps,
      extraProps = _objectWithoutPropertiesLoose(_ref, ["type", "requestKey", "selector", "defaultData", "multiple", "children", "component", "isDataEmpty", "showLoaderDuringRefetch", "noDataMessage", "errorComponent", "errorComponentProps", "loadingComponent", "loadingComponentProps"]);

  var query = useSelector(selector || getQuerySelector({
    type: type,
    requestKey: requestKey,
    defaultData: defaultData,
    multiple: multiple
  }));
  var dataEmpty = isDataEmpty(query);

  if (query.loading && (showLoaderDuringRefetch || dataEmpty)) {
    return LoadingComponent ? /*#__PURE__*/React.createElement(LoadingComponent, _extends({
      downloadProgress: query.downloadProgress,
      uploadProgress: query.uploadProgress
    }, loadingComponentProps)) : null;
  }

  if (query.error) {
    return ErrorComponent ? /*#__PURE__*/React.createElement(ErrorComponent, _extends({
      error: query.error
    }, errorComponentProps)) : null;
  }

  if (dataEmpty) {
    return noDataMessage;
  }

  if (children) {
    return children(query);
  }

  return /*#__PURE__*/React.createElement(Component, _extends({
    query: query
  }, extraProps));
};

Query.defaultProps = {
  isDataEmpty: function isDataEmpty(query) {
    return Array.isArray(query.data) ? query.data.length === 0 : !query.data;
  },
  showLoaderDuringRefetch: true,
  noDataMessage: null,
  multiple: false
};
Query.propTypes = process.env.NODE_ENV !== "production" ? {
  type: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  requestKey: PropTypes.string,
  selector: PropTypes.func,
  multiple: PropTypes.bool,
  defaultData: PropTypes.any,
  children: PropTypes.func,
  component: reactComponentPropType('Query'),
  isDataEmpty: PropTypes.func,
  showLoaderDuringRefetch: PropTypes.bool,
  noDataMessage: PropTypes.node,
  errorComponent: reactComponentPropType('Query'),
  errorComponentProps: PropTypes.objectOf(PropTypes.any),
  loadingComponent: reactComponentPropType('Query'),
  loadingComponentProps: PropTypes.objectOf(PropTypes.any)
} : {};
export default Query;
